src/osd/modules/input/input_evdev.cpp

```diff
@@ -0,0 +1,713 @@
+	{ BTN_FORWARD,		ITEM_ID_INVALID,	"FORWARD" },
+	{ BTN_BACK,		ITEM_ID_INVALID,	"BACK" },
+	{ BTN_TASK,		ITEM_ID_INVALID,	"TASK" },
+	{ BTN_TRIGGER,		ITEM_ID_INVALID,	"TRIGGER" },
+	{ BTN_THUMB,		ITEM_ID_INVALID,	"THUMB" },
+	{ BTN_THUMB2,		ITEM_ID_INVALID,	"THUMB2" },
+	{ BTN_TOP,		ITEM_ID_INVALID,	"TOP" },
+	{ BTN_TOP2,		ITEM_ID_INVALID,	"TOP2" },
+	{ BTN_PINKIE,		ITEM_ID_INVALID,	"PINKIE" },
+	{ BTN_BASE,		ITEM_ID_INVALID,	"BASE" },
+	{ BTN_BASE2,		ITEM_ID_INVALID,	"BASE2" },
+	{ BTN_BASE3,		ITEM_ID_INVALID,	"BASE3" },
+	{ BTN_BASE4,		ITEM_ID_INVALID,	"BASE4" },
+	{ BTN_BASE5,		ITEM_ID_INVALID,	"BASE5" },
+	{ BTN_BASE6,		ITEM_ID_INVALID,	"BASE6" },
+	{ BTN_DEAD,		ITEM_ID_INVALID,	"DEAD" },
+	{ BTN_SOUTH,		ITEM_ID_INVALID,	"A" },
+	{ BTN_EAST,		ITEM_ID_INVALID,	"B" },
+	{ BTN_C,		ITEM_ID_INVALID,	"C" },
+	{ BTN_NORTH,		ITEM_ID_INVALID,	"X" },
+	{ BTN_WEST,		ITEM_ID_INVALID,	"Y" },
+	{ BTN_Z,		ITEM_ID_INVALID,	"Z" },
+	{ BTN_TL,		ITEM_ID_INVALID,	"LEFT TRIGGER" },
+	{ BTN_TR,		ITEM_ID_INVALID,	"RIGHT TRIGGER" },
+	{ BTN_TL2,		ITEM_ID_INVALID,	"LEFT TRIGGER 2" },
+	{ BTN_TR2,		ITEM_ID_INVALID,	"RIGHT TRIGGER 2" },
+	{ BTN_SELECT,		ITEM_ID_INVALID,	"SELECT" },
+	{ BTN_START,		ITEM_ID_INVALID,	"START" },
+	{ BTN_MODE,		ITEM_ID_INVALID,	"MODE" },
+	{ BTN_THUMBL,		ITEM_ID_INVALID,	"LEFT THUMB" },
+	{ BTN_THUMBR,		ITEM_ID_INVALID,	"RIGHT THUMB" },
+	{ BTN_GEAR_DOWN,	ITEM_ID_INVALID,	"GEAR DOWN" },
+	{ BTN_GEAR_UP,		ITEM_ID_INVALID,	"GEAR UP" },
+	{ 0, ITEM_ID_INVALID, 0 },
+};
+
+idmap relaxis_map[] = {
+	{ REL_X,		ITEM_ID_XAXIS,		"X" },
+	{ REL_Y,		ITEM_ID_YAXIS,		"Y" },
+	{ REL_Z,		ITEM_ID_ZAXIS,		"Z" },
+	{ REL_RX,		ITEM_ID_RXAXIS,		"RX" },
+	{ REL_RY,		ITEM_ID_RYAXIS,		"RY" },
+	{ REL_RZ,		ITEM_ID_RZAXIS,		"RZ" },
+	{ REL_HWHEEL,		ITEM_ID_ADD_RELATIVE1,	"HWHEEL" },
+	{ REL_DIAL,		ITEM_ID_SLIDER2,	"DIAL" },
+	{ REL_WHEEL,		ITEM_ID_SLIDER1,	"WHEEL" },
+	{ REL_MISC,		ITEM_ID_ADD_RELATIVE2,	"EXTRA AXIS" },
+	{ REL_WHEEL_HI_RES,	ITEM_ID_SLIDER1,	"WHEEL" },
+	{ REL_HWHEEL_HI_RES,	ITEM_ID_ADD_RELATIVE1,	"HWHEEL" },
+	{ 0, ITEM_ID_INVALID, 0 },
+};
+
+idmap absaxis_map[] = {
+	{ ABS_X,		ITEM_ID_XAXIS,		"X" },
+	{ ABS_Y,		ITEM_ID_YAXIS,		"Y" },
+	{ ABS_Z,		ITEM_ID_ZAXIS,		"Z" },
+	{ ABS_RX,		ITEM_ID_RXAXIS,		"RX" },
+	{ ABS_RY,		ITEM_ID_RYAXIS,		"RY" },
+	{ ABS_RZ,		ITEM_ID_RZAXIS,		"RZ" },
+	{ ABS_THROTTLE,		ITEM_ID_SLIDER1,	"THROTTLE" },
+	{ ABS_RUDDER,		ITEM_ID_SLIDER2,	"RUDDER" },
+	{ ABS_WHEEL,		ITEM_ID_ADD_ABSOLUTE1,	"WHEEL" },
+	{ ABS_GAS,		ITEM_ID_ADD_ABSOLUTE2,	"GAS" },
+	{ ABS_BRAKE,		ITEM_ID_ADD_ABSOLUTE3,	"BRAKE" },
+	{ ABS_HAT0X,		ITEM_ID_ADD_ABSOLUTE4,	"HAT1 X" },
+	{ ABS_HAT0Y,		ITEM_ID_ADD_ABSOLUTE5,	"HAT1 Y" },
+	{ ABS_HAT1X,		ITEM_ID_ADD_ABSOLUTE6,	"HAT2 X" },
+	{ ABS_HAT1Y,		ITEM_ID_ADD_ABSOLUTE7,	"HAT2 Y" },
+	{ ABS_HAT2X,		ITEM_ID_ADD_ABSOLUTE8,	"HAT3 X" },
+	{ ABS_HAT2Y,		ITEM_ID_ADD_ABSOLUTE9,	"HAT3 Y" },
+	{ ABS_HAT3X,		ITEM_ID_ADD_ABSOLUTE10,	"HAT4 X" },
+	{ ABS_HAT3Y,		ITEM_ID_ADD_ABSOLUTE11,	"HAT4 Y" },
```

Comment:
People are going to expect hat switches to map to a set of digital inputs – you’re going to have to have to convert the “axis” value.

Reply:
Really?  It's fine with me in principle but I'd have expected hats to behave as axes not buttons (it turns out that I never used a device with a hat with mame so I suppose I don't start with an _actual_ expectation).

My reasoning: a hat behaves like a joystick because the directions are physically mutually excluded (you can never have "left" and "right" simultaneously, only either or none) so at first glance it makes no sense to map those to buttons (which you'd expect can be triggered in any combination).

If that's unconvincing, would it make sense to make that a configurable options instead?

Reply:
I thought our input code expected joysticks to show up as digital inputs too in any case?

Reply:
> I thought our input code expected joysticks to show up as digital inputs too in any case?

No, turning absolute axis values into “digital” toggle values happens in the core (look for `read_as_switch` members in src/emu/inputdev.cpp specifically).  The input module just reports axis values.